sockittome.js

```diff
@@ -15,3 +18,415 @@ module.exports = (io) => {
+
+//       socket.to(roomName).emit(SE_PLAY);
+//     });
+
+//     // Triggered when a user presses the pause button
+//     socket.on(SE_PAUSE, function () {
+//       console.log(
+//         `Client ${socket.id} in room ${roomName} emitted 'Pause' event.`
+//       );
+
+//       rooms[roomName].playInfo.isPlaying = false;
+
+//       socket.to(roomName).emit(SE_PAUSE);
+//     });
+
+//     // Triggered when a user presses the "next" button
+//     socket.on(SE_NEXT, function () {
+//       console.log(
+//         `Client ${socket.id} in room ${roomName} emitted 'Next' event.`
+//       );
+
+//       rooms[roomName].playInfo.currentPlayingIndex++;
+//       if (
+//         rooms[roomName].playInfo.currentPlayingIndex >
+//         rooms[roomName].playInfo.playlist.length - 1
+//       )
+//         rooms[roomName].playInfo.currentPlayingIndex =
+//           rooms[roomName].playInfo.playlist.length - 1;
+
+//       socket.to(roomName).emit(SE_NEXT);
+//     });
+
+//     // Triggered when a user presses the "previous" button
+//     socket.on(SE_PREV, function () {
+//       console.log(
+//         `Client ${socket.id} in room ${roomName} emitted 'Prev' event.`
+//       );
+
+//       rooms[roomName].playInfo.currentPlayingIndex--;
+//       if (rooms[roomName].playInfo.currentPlayingIndex < 0)
+//         rooms[roomName].playInfo.currentPlayingIndex = 0;
+
+//       socket.to(roomName).emit(SE_PREV);
+//     });
+
+//     // Triggered when a user presses the "Choose Song" button
+//     socket.on(SE_ADD_SONG_TO_PLAYLIST, function (songUrl) {
+//       console.log(
+//         `Client ${socket.id} in room ${roomName} emitted 'addSongToPlaylist' event with ${songUrl}`
+//       );
+
+//       rooms[roomName].playInfo.playlist.push(songUrl);
+
+//       socket.to(roomName).emit(SE_ADD_SONG_TO_PLAYLIST, songUrl);
+//     });
+
+//     // Triggered when a user presses the "Choose Song" button
+//     socket.on(SE_ADD_SONG_TO_QUEUE, function (queueSong) {
+//       console.log(
+//         `Client ${socket.id} in room ${roomName} emitted 'addSongToQueue' event with ${queueSong["title"]}`
+//       );
+
+//       rooms[roomName].playInfo.queue.push(queueSong);
+
+//       socket.to(roomName).emit(SE_ADD_SONG_TO_QUEUE, queueSong);
+//     });
+//   });
+// };
+
+
+// Redux Reducers
+
+let lastId = 0;
+let songIndex = 0;
+
+const CREATE_ROOM = "createRoom";
+const SONG_ADDED = "songAdded";
+const PLAY_PRESSED = "playPressed";
+const PAUSE_PRESSED = "pausePressed";
+const PREV_PRESSED = "prevPressed";
+const NEXT_PRESSED = "nextPressed";
+
+export default function playlistReducer(state = [], action) {
+  switch (action.type) {
+    case "CREATE_ROOM":
+      return state.concat([
+        {
+          roomId: ++lastId,
+          socketUserMap: action.payload.user,
+          playlist: [],
+          queue: [],
+          currentPlayingIndex: 0,
+          currentPlayingTime: 0,
+          isPlaying: false,
+        }
+      ]);
```

Comment:
This is a good use case for [normalization](https://github.com/paularmstrong/normalizr#quick-start) due to arrays being tough to mutate / lookup values from.

```tsx
let nextRoomId = 0;

export default function playlistReducer(state = {}, action) {
  switch (action.type) {
    case "CREATE_ROOM": {
     const roomId =  ++nextRoomId;
      return {
         ...state,
         [roomId]:  {
          roomId,
          socketUserMap: action.payload.user,
          playlist: [],
          queue: [],
          currentPlayingIndex: 0,
          currentPlayingTime: 0,
          isPlaying: false,
       }
   }
}
```

Normalization helps with both getting and setting of rooms (especially if we only ever are acting on / using 1!) see how updates become simpler:

```tsx
 case "PLAY_PRESSED": {
      return  {
        ...state, // keep all other keys the same
        [action.payload.id]: {
           ...state[action.payload.id], // keep the rest of the room state the same
           isPlaying: true
       }
   }
}
```

Reply:
Fixed!